app/tests/spec/lib/app-start.js

```diff
@@ -159,6 +173,131 @@ function (chai, sinon, AppStart, Session, Constants, p,
           });
       });
     });
+
+    describe('initializeAuthenticationBroker', function () {
+      function testExpectedBrokerCreated(ExpectedBroker) {
+        return appStart.initializeAuthenticationBroker()
+          .then(function () {
+            assert.isTrue(
+              appStart._authenticationBroker instanceof ExpectedBroker);
+          });
+      }
+
+      beforeEach(function () {
+        appStart = new AppStart({
+          window: windowMock,
+          router: routerMock,
+          history: historyMock,
+          user: userMock
+        });
+      });
+
+      describe('fx-desktop', function () {
+        it('returns an FxDesktop broker if `service=sync`', function () {
+          windowMock.location.search = Url.objToSearchString({
+            service: Constants.FX_DESKTOP_SYNC
+          });
+
+          return testExpectedBrokerCreated(FxDesktopBroker);
+        });
+
+        it('returns an FxDesktop broker if `context=fx_desktop_v1`', function () {
+          windowMock.location.search = Url.objToSearchString({
+            context: Constants.FX_DESKTOP_CONTEXT
+          });
+
+          return testExpectedBrokerCreated(FxDesktopBroker);
+        });
+      });
+
+      describe('web channel', function () {
+        it('returns a WebChannel broker if `webChannelId` is present', function () {
+          windowMock.location.search = Url.objToSearchString({
+            webChannelId: 'channel id'
+          });
+
+          return testExpectedBrokerCreated(WebChannelBroker);
+        });
+
+        it('returns a WebChannel broker if verifying in the same brower where a signup was initiated from a web channel', function () {
+          Session.set('oauth', {
+            //jshint camelcase: false
+            client_id: 'client id',
+            webChannelId: 'channel id'
+          });
+
+          windowMock.location.search = Url.objToSearchString({
+            code: 'code',
+            service: 'client id'
+          });
+
+          return testExpectedBrokerCreated(WebChannelBroker);
+        });
+      });
+
+      describe('iframe', function () {
+        it('returns an Iframe broker if `context=iframe` is present and in an iframe', function () {
+          windowMock.top = new WindowMock();
+          windowMock.location.search = Url.objToSearchString({
+            context: Constants.IFRAME_CONTEXT,
+            //jshint camelcase: false
+            client_id: 'client id'
+          });
+
+          return testExpectedBrokerCreated(IframeBroker);
+        });
+
+        it('returns a Redirect broker if `context=iframe` is not present and in an iframe', function () {
+          windowMock.top = new WindowMock();
+          windowMock.location.search = Url.objToSearchString({
+            //jshint camelcase: false
+            client_id: 'client id'
+          });
```

Comment:
I wonder if we should error out in this case. The relier has asked for an iframe, but there's not in an iframe, so this is probably not going to end well...

Given the importance of this PR to marketplace, we can consider this in a follow-up if you prefer.